Support for long running commands

gerbyzation opened the pull request:
In response to #7, add support to test long running commands.

Looking forward to hear what you think it is! If you'd be interested to accept this there is probably a few things that can be improved and tested. 

fix #7

rendon commented:
Hello @gerbyzation, thanks for your interest in the package. The idea sounds good, I'll take a closer look and review the code as soon as I get some chance, most likely this weekend.

rendon commented:
My first comment is on the commit messages. Could you follow this guide https://chris.beams.io/posts/git-commit/?

rendon commented on main.go:
```diff
@@ -17,34 +18,58 @@ import (
-	executed  bool
-	stdout    string
-	stderr    string
+	status    string
```
The `executed` flag was simple because all we cared about was to know if a command  had been executed, or not. With long-running programs, we have a few more states. Off the top of my head:
- initialized
- started
- running
- finished

I would think about all the possible states and how they can be applied to both long-running and short-lived commands.

gerbyzation replied:
I think started can be derived from the state being either running or finished, which would leave us with three types of state that can be recorded on `Command`:
 * initialized
 * running - only relevant for long running commands. 
 * finished

rendon commented on main.go:
```diff
@@ -17,34 +18,58 @@ import (
 
-func (c *Cmd) validate() {
-	if !c.executed {
+func (c *Cmd) validateIsDone() {
+	if c.status != "executed" {
+		log.Fatal(ErrCmdNotFinished)
```
Based on my other comment about all the possible states, I would suggest to use "Finished" for the status instead of "executed".

rendon commented on main.go:
```diff
@@ -82,9 +107,70 @@ func (c *Cmd) Run() {
+	go func() {
+		scanner := bufio.NewScanner(stdoutPipe)
+		for scanner.Scan() {
+			c.stdout.mu.Lock()
+			c.stdout.content += scanner.Text()
+			c.stdout.mu.Unlock()
```
I'm curious to know what type of concurrency issues you ran into.

gerbyzation replied:
I hope https://github.com/rendon/testcli/pull/8#discussion_r579804006 explains this well enough.

rendon commented on main.go:
```diff
@@ -82,9 +107,70 @@ func (c *Cmd) Run() {
+
+// Wait waits for the command to exit
+func (c *Cmd) Wait() {
+	c.validateHasStarted()
```
Let's see. `validateHasStarted()` checks if status == "initialized". However, after running `Start()` the status will be "running". I am assuming that `Wait()` runs after `Start()`, is that not the case?

gerbyzation replied:
Yes, `validateHasStarted` will raise an error if `status == "initialized". `Wait()` should be called after `Start()`. `Start()` will advance `status` to `running`, so this will only raise if it's called before `Start()`. I'll clarify the comment and add a test.

gerbyzation replied:
We'll need to move away from `log.Fatal` before this can be tested.

rendon commented on main.go:
```diff
@@ -96,7 +182,7 @@ func Run(name string, arg ...string) {
 // Error is the command's error, if any.
 func (c *Cmd) Error() error {
-	c.validate()
+	c.validateIsDone()
```
I'm having second thoughts about these validation methods, I'm no longer sure it's a good idea to halt the execution (`log.Fatal()`), it might be better to return an error and let the caller decide what to do. For example:
```go
if !c.hasFinished() {
	log.Printf("The program hasn't started or is still running...")
	// Should I halt the program? Or should I return a special error?
}
````

gerbyzation replied:
Yes, I think `log.Fatal` is not a great idea. Did some tests and when this is called it stops the rest of the test suite from running. Returning an error could work, but would make tests a bit more verbose. Would it be an idea to pass `*testing.T` down to the command so we can replace the `log.Fatal` calls with `t.Fatal`?

rendon commented on main.go:
```diff
@@ -107,8 +193,10 @@ func Error() error {
-	c.validate()
-	return c.stdout
+	c.validateHasStarted()
+	c.stdout.mu.Lock()
+	defer c.stdout.mu.Unlock()
+	return c.stdout.content
```
Do you need a lock for read operations? What is the problem you're trying to prevent?

gerbyzation replied:
Initially I hadn't added the `retryStringTest` funcyet  and needed to prevent reading the output content while this was still being written to prevent race conditions and unnecessary failures. This seemed to work initially but still ocasionally gave failing tests. 

I decided to add a retry mechanism, as my current approach didn't seem fully reliable. In addition the mutex approach would not work with long running commands that might need longer to reply with the desired output. 

I'm not sure if it's worth keeping around with the retry mechanism, at best it prevents an extra retry attempt because it forces the write to finish before executing the test.

rendon commented on main.go:
```diff
@@ -193,12 +287,34 @@ func StdoutMatches(regex string) bool {
+
+// retryStringTest takes in a testFunc and will test output for the expected string until either it
+// finds the expected string or times out (default 1 second)
+func retryStringTest(testFunc func(string, string) bool, output *output, expected string) bool {
```
What is the use case for this method?

gerbyzation replied:
As mentioned in the comment on the mutex I added a retry mechanism in the `*Contains` and `*Matches` methods to prevent race conditions as the output sometimes seemed to be delayed. It also expands the usecase to long running commands/processes that are delayed in producing the desired output. The `testFunc` can be something like `strings.Contains`, and will be executed with the content of the `output` and `expected` string every 100ms until either a match is found or it reaches the timeout. 

At the moment the timeout  (and interval) is hardcoded but this should be configurable I suppose so it can be tailored to specific usecases

rendon reviewed (commented):
Thanks for the work you put in. I have a few questions about the implementation.

gerbyzation commented on main.go:
```diff
@@ -17,34 +18,58 @@ import (
+	"time"
 )
 
+type output struct {
+	content string
+	mu      *sync.Mutex
+}
```
As explained in https://github.com/rendon/testcli/pull/8#discussion_r579804006 I'm not sure if we should keep this or not. If decided it stays I can add some documentation to this ofcourse

gerbyzation commented on main.go:
```diff
@@ -17,34 +18,58 @@ import (
+	}
+}
+
+func (c *Cmd) validateHasStarted() {
+	if c.status == "initialized" {
```
Yes, so a process that has started can be running or finished. The one state that would be invalid is `initialized`. This checks if it's in the one state that would be invalid and if yes; raising an error.

Need to think if we can write this in a more descriptive way or else add a comment to clarify .

gerbyzation commented:
@rendon thanks for the pointer on commit messages. Do you want me to rewrite the history or just for going forward?

rendon commented:
> @rendon thanks for the pointer on commit messages. I'll take it on board, if you ways to still improve later on please do let me know! Do you want me to rewrite the history or just for going forward?

You can update the commit messages by running `git rebase` in interactive mode. For example:
```sh
git rebase -i HEAD~5 # To edit the last 5 commits
```

gerbyzation commented:
@rendon Yeah I'm aware, just wanted to make sure you're OK with me force pushing as not everyone is as keen on that 🙂 I've updated my commits, hope this is a step in the right direction. 

I've also switched to using `t.Fatal`. Found it to be a good improvement. Not what your thoughts on this are or if you'd want this in the same PR. Would be great to get your thoughts on this and the review discussions.

gerbyzation commented:
I forgot to ask, does the whole test suite pass for you on master? If so, what environment do you develop in? I'm working on macOS 10.15 and these tests fail for me. 

```
➜  testcli git:(long-running) go test .
--- FAIL: TestSetEnv (0.01s)
    main_test.go:20: -n bar

    main_test.go:21: stdout failed to include input
--- FAIL: TestStderr (1.01s)
    main_test.go:112: Expected "usage: cp [-R [-H | -L | -P]] [-fi | -n] [-apvXc] source_file target_file\n       cp [-R [-H | -L | -P]] [-fi | -n] [-apvXc] source_file ... target_directory\n" to contains "MISSING"
--- FAIL: TestPackageStderr (1.00s)
    main_test.go:122: Expected "usage: cp [-R [-H | -L | -P]] [-fi | -n] [-apvXc] source_file target_file\n       cp [-R [-H | -L | -P]] [-fi | -n] [-apvXc] source_file ... target_directory\n" to contains "missing"
--- FAIL: TestStderrMatches (2.01s)
    main_test.go:168: Expected "usage: cp [-R [-H | -L | -P]] [-fi | -n] [-apvXc] source_file target_file\n       cp [-R [-H | -L | -P]] [-fi | -n] [-apvXc] source_file ... target_directory\n" to match "cp.*operand"
FAIL
FAIL	github.com/rendon/testcli	10.639s
FAIL
```

rendon commented:
Apologies for the delay. Yes, the tests pass for me on Linux (Ubuntu):
```
$ go test
PASS
ok      github.com/rendon/testcli       0.029s
```

The last commit I have in master is `aebbc9548fac6fa6b0130b33c9722238cdc9f1ae`.

rendon commented on main.go:
```diff
@@ -9,43 +9,79 @@
+
+const (
+	// INITIALIZED represents the state of Command before it's started with Run() or Start()
+	INITIALIZED = "initialized"
+	// RUNNING represents the state of Command while it's running
+	RUNNING = "running"
+	// FINISHED represents the state of Command after it has exited successfully or not
+	FINISHED = "finished"
```
Please use Camel Case notation for variables and constants. See https://github.com/golang/go/wiki/CodeReviewComments#mixed-caps.

gerbyzation replied:
Ah, didn't know golang had different standard for constant naming, will get that updated!

rendon commented:
The implementation seems good overall, just make sure we don't break anything. I'll try to squeeze some time these days to see the new feature in action.

gerbyzation commented:
@randon ok, must be some small differences in the CLI tools. Have updated the const names to camelCase

Define breaking? It would be a breaking change if that's what you're asking - `t.Testing` needs to be passed in from the test. If that should live in a separate release/feature branch I'm happy to revert it. Overall I'd be in favour of the change, as at the moment the test runner is killed by `log.Fatal` when it encounters the first failing test. 

If breaking the API to fix broken functionality is not an option even in a more significant release I wonder if it would be possible to come up with new functions for `Command` and `Run` that suppor the `t` argument (not sure yet on good names). For the existing functions I imagine we could create a stub for `t.Testing` with a no-op for the `Helper` method and `Fatal` would wrap `log.Fatal`.

rendon commented:
@gerbyzation By breaking I mean not breaking existing users. However, recently I created a release for the tool, users should start using releases, which will be stable, and not consume mainline directly.

I'll take a final look this weekend and merge the changes.
